pyro/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,144 @@
+                                    loc.size()))
+            self.sigma = covariance_matrix
+            self.sigma_cholesky = potrf_compat(covariance_matrix)
+        else:
+            if not scale_tril.dim() == 2:
+                raise ValueError("The Cholesky decomposition of the covariance matrix must be a matrix, "
+                                 "but got scale_tril.size() = {}".format(
+                                    loc.size()))
+            self.sigma = torch.mm(scale_tril.transpose(0, 1), scale_tril)
+            self.sigma_cholesky = scale_tril
+        if loc.dim() > 1:
+            raise ValueError("The mean must be a vector, but got loc.size() = {}".format(loc.size()))
+
+        super(MultivariateNormal, self).__init__(*args, **kwargs)
+
+    def batch_shape(self, x=None):
+        mu = self.mu.expand(self.batch_size, *self.mu.size()).squeeze(0)
+        if x is not None:
+            if x.size()[-1] != mu.size()[-1]:
+                raise ValueError("The event size for the data and distribution parameters must match.\n"
+                                 "Expected x.size()[-1] == self.mu.size()[0], but got {} vs {}".format(
+                                    x.size(-1), mu.size(-1)))
+            try:
+                mu = mu.expand_as(x)
+            except RuntimeError as e:
+                raise ValueError("Parameter `mu` with shape {} is not broadcastable to "
+                                 "the data shape {}. \nError: {}".format(mu.size(), x.size(), str(e)))
+
+        return mu.size()[:-1]
+
+    def event_shape(self):
+        return self.mu.size()[-1:]
+
+    def sample(self):
+        """
+        Generate a sample with the specified covariance matrix and mean.
+        Differentiation wrt. to the covariance matrix is only supported on PyTorch version 0.3.0 or higher.
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.sample`
+
+        """
+        batch_size = self.batch_size
+        uncorrelated_standard_sample = Variable(torch.randn(batch_size, *self.mu.size()).type_as(self.mu.data))
+        transformed_sample = self.mu + torch.mm(uncorrelated_standard_sample, self.sigma_cholesky)
+        return transformed_sample if self.reparameterized else transformed_sample.detach()
+
+    def batch_log_pdf(self, x):
+        if not self.normalized and self.sigma_cholesky.requires_grad:
+            warnings.warn("Gradients will not take normalization into account if normalized=False.")
+
+        batch_size = x.size()[0] if len(x.size()) > len(self.mu.size()) else 1
+        batch_log_pdf_shape = self.batch_shape(x) + (1,)
+        x = x.view(batch_size, *self.mu.size())
+        normalization_factor = torch.log(
+            self.sigma_cholesky.diag()).sum() + (self.mu.size()[0] / 2) * np.log(2 * np.pi) if self.normalized else 0
+        sigma_inverse = matrix_inverse_compat(self.sigma, self.sigma_cholesky)
```

Comment:
Your review comment was helpful in understanding how we'll update this code in the future. Could you insert that comment here as a `# TODO It may be useful to switch between matrix_inverse_compat() and linear_solve_compat() ...`?